@joeyAghion : initial lib implementation

cavvia opened the pull request:
This extracts our `Searchable` module from Gravity, removing the `Mongoid` and `Sidekiq` dependencies to create a more generic Elasticsearch indexing/searching library. 

Three main `Stella` modules are extracted: `Searchable`, `Analysis` and `Helpers`. These remain `ActiveSupport::Concern` modules. A `Stella::Query` base object is provided which we will extend in Gravity with some of our specific filters and aggregations (such as `visible_to_public`).

The `#stella_search` method is the search interface provided, which leverages `Stella::Searchable` specific indexing options when searching models. It defines a `multi_match` search across fields, with custom weightings and boosts applied by default. I didn't want to override the `elasticsearch-model` `#search` method.

Gravity will also need to override the inline indexing callbacks defined here on `after_save` and `after_destroy` events, namely `#es_index` and `#es_delete`, and replace them with Sidekiq calls.

A Gravity PR will follow. 

Note the repo is public and intended as an open source project.

cc @mzikherman as this contains some of your `FilteredQuery` code from Gravity (in the `Stella::Query` base class).

cavvia committed:
initial stella libs and specs work

cavvia committed:
readme v1

cavvia committed:
add raw option

cavvia committed:
update readme and doc

cavvia committed:
rspec conf

cavvia committed:
tests for filters and slug fields

cavvia committed:
readme fixes

cavvia committed:
gemspec

cavvia committed:
more readme fixes

cavvia commented on Gemfile:
```diff
@@ -0,0 +1,15 @@
+ruby '2.2.2'
+
+gem 'activesupport', '~> 4.2.2'
+gem 'elasticsearch-model'
```
I would like to pin this but the last tag is from May 2016, so i'd need to double-check we weren't relying on any features developed in the meantime (there has been on-going development). I'll try to pin the gems below this though.

cavvia commented on lib/stella/helpers.rb:
```diff
@@ -0,0 +1,103 @@
+    @@types = []
+
+    included do
+      if defined? Rails
```
I'm not sure how best to deal with this Rails dependency. Environment versioning on index names seems quite important for the lib to include but I guess it's not a hard requirement - the host repo could override the index naming scheme with that perhaps.

joeyAghion replied:
I've seen gems provide optional Rails-friendly integration via a `railtie.rb`. I'm not sure if the convention has changed significantly in recent Rails versions. (In any case, I think this sort of `defined?` check is safe because `Rails` will always be loaded first in that context.)

Same applies to the check above in `analysis.rb`.

cavvia replied:
I'm going to remove the `env` from the naming scheme altogether and just provide the override hook to client app. I will update the readme with info on that.

cavvia commented on lib/stella/helpers.rb:
```diff
@@ -0,0 +1,103 @@
+        env = 'test'
+      end
+
+      index_name [search_index_name, env].join('_')
```
I'll refactor this to add a hook which will allow for override of the naming scheme.

cavvia replied:
Should also probably support some kind of APP_NAME prefix for the index name.

cavvia commented on lib/stella/query.rb:
```diff
@@ -0,0 +1,137 @@
+      { bool: { must: query[:filter][:bool][:must] } }
+    end
+
+    def add_aggregation_query(type, filter, aggregation)
```
I might remove aggregation support from here to leave it as a Gravity override.

joeyAghion commented on README.md:
````diff
@@ -1,2 +1,103 @@
+The module will try to use Elasticsearch on `localhost:9200` by default. You can configure your global ES client like so:
+
+```ruby
+Elasticsearch::Model.client = Elasticsearch::Client.new host: myhost, log: true
````
It might be helpful to provide a real string in place of `myhost` to make the format expectations clear.

cavvia replied:
Yes will fix this.

joeyAghion commented on README.md:
```diff
@@ -1,2 +1,103 @@
+
+For further search customization, see the [elasticsearch dsl](https://github.com/elastic/elasticsearch-rails/tree/master/elasticsearch-model#the-elasticsearch-dsl). 
+
+Stella works with any ActiveRecord compatible database (MySQL, sqlite, Postgres) as well as MongoDB.
```
But it presumably doesn't depend directly on the database, so what's the actual dependency? Activemodel? Activerecord or Mongoid?

joeyAghion replied:
Also, this README would benefit from a copyright and an explicit license (probably MIT).

cavvia replied:
The dependency is ActiveRecord or Mongoid. I'll clarify that.

cavvia replied:
The license is in a separate file as per github convention, I'll include a copyright and link to the license at the bottom.

joeyAghion commented on Rakefile:
```diff
@@ -0,0 +1,7 @@
+
+desc "Run all tests"
+RSpec::Core::RakeTask.new(:spec) do |spec|
+  spec.pattern = "spec/*_spec.rb"
```
Seems like this would be the default pattern.

joeyAghion commented on lib/stella/helpers.rb:
```diff
@@ -0,0 +1,103 @@
+      end
+
+      def bulk_index(batch_of_ids)
+        __elasticsearch__.client = Elasticsearch::Client.new url: ENV['ELASTICSEARCH_URL']
```
Should this ENV value be configured elsewhere rather than reach from the gem directly into the environment?

cavvia replied:
Hmm. There's an argument for keeping client configuration out of the library altogether. I'm not sure why we need this line in any case - this could be legacy code left over from our cluster migration period.

joeyAghion commented on stella.gemspec:
```diff
@@ -0,0 +1,29 @@
+
+Gem::Specification.new do |gem|
+  gem.name = "stella"
+  gem.homepage = "http://github.com/artsy/stella"
```
`https`

joeyAghion commented on stella.gemspec:
```diff
@@ -0,0 +1,29 @@
+
+  gem.description = <<-EOS
+    Make your Ruby objects searchable with Elasticsearch.
+  EOS
```
The `<<-EOS` doesn't seem necessary here.

joeyAghion commented on Gemfile:
```diff
@@ -0,0 +1,15 @@
+gem 'elasticsearch-model'
+gem 'activemodel'
+gem 'typhoeus'
+gem 'activerecord'
```
I have questions about some of these dependencies.
* If you're using `activesupport` only for the `ActiveSupport::Concern` conventions/syntax, it might be simpler to implement that internally.
* I don't see direct usage of Typhoeus.
* I don't think activerecord should be a dependency. ~We certainly wouldn't want to bring it into Gravity.~ (Just realized it's already there! WHA?!) If you need it for tests, it can be added in the .gemspec as a development dependency (see comment below). And the callbacks that I think _are_ depended on come from ActiveModel I believe.
* If you somehow _do_ need activerecord, I don't think `activemodel` is required since it's already a transitive dependency.

cavvia replied:
* I think I can move `ActiveRecord` to a dev dependency as you suggest and keep `ActiveModel` as I think I only need to reference the `after_save` and `after_delete` callbacks.
* The [elasticsearch-transport](https://github.com/elastic/elasticsearch-ruby/tree/master/elasticsearch-transport) gem, used under the hood, pulls in `Faraday` and detects the presence of Typhoeus automatically, all that it needs is a `require 'typhoeus'; require 'typhoeus/adapters/faraday'` in order to use Typhoeus for transport, so I would need to add that `require` in order to make this dependency useful. The question is, should we add support for Typhoeus here or leave it upto the client app? I'm probably in favour of the former but it does add an extra dependency.

joeyAghion replied:
Ah... I see the mention of that in the [elasticsearch-transport doc](https://github.com/elastic/elasticsearch-ruby/tree/master/elasticsearch-transport). Since this doesn't use Typhoeus directly, I think ideally it should allow multiple adapters the same way that gem does. It might be worth "recommending" that Typhoeus (or another library supporting persistent connections) be required in this project's README.

joeyAghion commented on Gemfile:
```diff
@@ -0,0 +1,15 @@
+  gem 'elasticsearch-extensions'
+  gem 'rspec-expectations'
+  gem 'sqlite3'
+end
```
I understood that gem dependencies should be declared primarily in the `.gemspec` file, with only loose version constraints (so as not to conflict with dependencies of the including apps). The Gemfile would then contain just a simple `gemspec` declaration referring to that file.

http://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/

These test dependencies would become "development" dependencies in that scheme.

cavvia replied:
Oh yes, you're right, will switch to that style.

joeyAghion commented on Gemfile:
```diff
@@ -0,0 +1,15 @@
+ruby '2.2.2'
```
I think gems aren't supposed to declare Ruby versions, leaving that to the including apps. (Many are intentionally tested against multiple versions.)

cavvia replied:
Yes makes sense.

joeyAghion commented on Gemfile.lock:
```diff
@@ -0,0 +1,138 @@
+   ruby 2.2.2p95
+
+BUNDLED WITH
+   1.12.5
```
Also the Gemfile.lock shouldn't be checked in for a gem.

cavvia replied:
:+1:

cavvia committed:
fixes dependencies and readme issues

cavvia committed:
gitignore file

cavvia committed:
remove Rails reference and add doc on custom index naming

cavvia committed:
remove Rakefile

cavvia committed:
add rubocop and travis support, remove typhoeus dependency

cavvia committed:
remove aggregation code

cavvia commented:
I think I've addressed everything here - 

* I removed the use of `Rails.env` in the index naming scheme, instead adding a line of doc around how to override the naming scheme (which is recommended). 
* I removed the Typhoeus dependency - `elasticsearch-ruby` already has doc on that.
* I moved the `ActiveRecord` dependency to development.
* Moved all deps management to the gemspec.
* Added rubocop and travis support.
* Removed the aggregations code.
* Updated the README accordingly.

joeyAghion commented on lib/stella/helpers.rb:
```diff
@@ -0,0 +1,96 @@
+require 'typhoeus/adapters/faraday'
```
Should this be removed now that `typhoeus` isn't explicitly required?

cavvia replied:
Good spot! Travis would have picked this up. Fixed now.

cavvia committed:
remove typhoeus dependency
